File: NeetCode/Math_And_Geometry/Insert_GCD_Linked_List/insert__g_c_d__linked__list.py
from typing import Optional, List
import heapq
from collections import defaultdict
from common.listnode import ListNode, build_list, print_list
import math


class Solution:
    def insertGreatestCommonDivisors(
        self, head: Optional[ListNode]
    ) -> Optional[ListNode]:

        ptr = head

        while ptr and ptr.next:
            g = math.gcd(ptr.val, ptr.next.val)
            tmp = ptr.next
            ptr.next = ListNode(g, next=tmp)
            ptr = ptr.next.next

        return head

        pass


if __name__ == "__main__":
    sol = Solution()
    head = [18, 6, 10, 3]
    head = build_list(head)
    print_list(sol.insertGreatestCommonDivisors(head))
    print("Running Solution...")

File: NeetCode/Math_And_Geometry/Insert_GCD_Linked_List/Insert_GCD_Linked_List.cpp
// Insert_GCD_Linked_List.cpp
//
// Compile locally with:
// clang++ Insert_GCD_Linked_List.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *insertGreatestCommonDivisors(ListNode *head)
    {
        ListNode *ptr = head;
        while (ptr && ptr->next)
        {
            ListNode *tmp = ptr->next;
            ptr->next = new ListNode(gcd(ptr->val, ptr->next->val), tmp);
            ptr = ptr->next->next;
        }
        return head;
    }
};

int main()
{
    cout << "Running Insert_GCD_Linked_List.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
